rueidisprob/slidingbloomfilter.go

```diff
@@ -0,0 +1,438 @@
+	if err != nil {
+		return fmt.Errorf("failed to check rotation: %w", err)
+	}
+
+	buf := bytesPool.Get(0, len(keys)*int(s.hashIterations)*8)
+	defer bytesPool.Put(buf)
+
+	indexes := s.indexes(keys, &buf.s)
+
+	args := make([]string, 0, len(indexes)+1)
+	args = append(args, s.hashIterationString)
+	args = append(args, indexes...)
+
+	resp := s.addMultiScript.Exec(ctx, s.client, s.addMultiKeys, args)
+	return resp.Error()
+}
+
+func (s *slidingBloomFilter) indexes(keys []string, buf *[]byte) []string {
+	allIndexes := make([]string, 0, len(keys)*int(s.hashIterations))
+	size := uint64(s.size)
+
+	for _, key := range keys {
+		h1, h2 := hash([]byte(key))
+		for i := uint(0); i < s.hashIterations; i++ {
+			offset := len(*buf)
+			*buf = strconv.AppendUint(*buf, index(h1, h2, i, size), 10)
+			allIndexes = append(allIndexes, rueidis.BinaryString((*buf)[offset:]))
+		}
+	}
+	return allIndexes
+}
+
+func (s *slidingBloomFilter) Exists(ctx context.Context, key string) (bool, error) {
+	exists, err := s.ExistsMulti(ctx, []string{key})
+	if err != nil {
+		return false, err
+	}
+
+	return exists[0], nil
+}
+
+func (s *slidingBloomFilter) ExistsMulti(ctx context.Context, keys []string) ([]bool, error) {
+	if len(keys) == 0 {
+		return nil, nil
+	}
+
+	err := s.checkAndRotate(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check rotation: %w", err)
+	}
+
+	buf := bytesPool.Get(0, len(keys)*int(s.hashIterations)*8)
+	defer bytesPool.Put(buf)
+
+	indexes := s.indexes(keys, &buf.s)
+
+	args := make([]string, 0, len(indexes)+1)
+	args = append(args, s.hashIterationString)
+	args = append(args, indexes...)
+
+	resp := s.existsMultiScript.Exec(ctx, s.client, s.existsMultiKeys, args)
+	if resp.Error() != nil {
+		return nil, resp.Error()
+	}
+
+	arr, err := resp.ToArray()
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]bool, len(keys))
+	for i, el := range arr {
+		v, err := el.AsBool()
+		if err != nil {
+			if rueidis.IsRedisNil(err) {
+				result[i] = false
+				continue
+			}
+
+			return nil, err
+		}
+
+		result[i] = v
+	}
+	return result, nil
+}
+
+func (s *slidingBloomFilter) Reset(ctx context.Context) error {
+	resp := s.client.Do(ctx,
+		s.client.B().
+			Eval().
+			Script(slidingBloomFilterResetScript).
+			Numkeys(4).
+			Key(s.addMultiKeys...).
+			Build(),
+	)
+	return resp.Error()
+}
+
+func (s *slidingBloomFilter) Delete(ctx context.Context) error {
+	resp := s.client.Do(ctx, s.client.B().Del().Key(s.rotateKeys...).Build())
+	return resp.Error()
+}
+
+func (s *slidingBloomFilter) Count(ctx context.Context) (uint64, error) {
+	resp := s.client.Do(
+		ctx,
+		s.client.B().
+			Get().
+			Key(s.counter).
+			Build(),
+	)
+	count, err := resp.AsUint64()
+	if err != nil {
+		if rueidis.IsRedisNil(err) {
+			return 0, nil
+		}
+
+		return 0, err
+	}
+
+	return count, nil
+}
+
+func (s *slidingBloomFilter) checkAndRotate(ctx context.Context) error {
+	// Fast path - check if enough time has passed without locking
```

Comment:
Accessing without locking is incorrect. You still need to use lock here, otherwise it is still data race.

Reply:
That don't sound bad,  I’ll explore that idea